crates/typst/src/layout/grid/mod.rs

```diff
@@ -0,0 +1,370 @@
+    #[borrowed]
+    pub rows: TrackSizings,
+
+    /// The gaps between rows & columns.
+    ///
+    /// If there are more gutters than defined sizes, the last gutter is repeated.
+    #[external]
+    pub gutter: TrackSizings,
+
+    /// The gaps between columns. Takes precedence over `gutter`.
+    #[parse(
+        let gutter = args.named("gutter")?;
+        args.named("column-gutter")?.or_else(|| gutter.clone())
+    )]
+    #[borrowed]
+    pub column_gutter: TrackSizings,
+
+    /// The gaps between rows. Takes precedence over `gutter`.
+    #[parse(args.named("row-gutter")?.or_else(|| gutter.clone()))]
+    #[borrowed]
+    pub row_gutter: TrackSizings,
+
+    /// How to fill the cells.
+    ///
+    /// This can be a color or a function that returns a color. The function is
+    /// passed the cells' column and row index, starting at zero. This can be
+    /// used to implement striped grids.
+    ///
+    /// ```example
+    /// #grid(
+    ///   fill: (col, row) => if calc.even(col + row) { luma(240) } else { white },
+    ///   align: center + horizon,
+    ///   columns: 4,
+    ///   [X], [O], [X], [O],
+    ///   [O], [X], [O], [X],
+    ///   [X], [O], [X], [O],
+    ///   [O], [X], [O], [X]
+    /// )
+    /// ```
+    #[borrowed]
+    pub fill: Celled<Option<Paint>>,
+
+    /// How to align the cells' content.
+    ///
+    /// This can either be a single alignment, an array of alignments
+    /// (corresponding to each column) or a function that returns an alignment.
+    /// The function is passed the cells' column and row index, starting at zero.
+    /// If set to `{auto}`, the outer alignment is used.
+    ///
+    /// ```example
+    /// #grid(
+    ///   columns: 3,
+    ///   align: (x, y) => (left, center, right).at(x),
+    ///   [Hello], [Hello], [Hello],
+    ///   [A], [B], [C],
+    /// )
+    /// ```
+    #[borrowed]
+    pub align: Celled<Smart<Align>>,
+
+    /// How to [stroke]($stroke) the cells.
+    ///
+    /// Grids have no strokes by default, which can be changed by setting this
+    /// option to the desired stroke.
+    ///
+    /// _Note:_ Richer stroke customization for individual cells is not yet
+    /// implemented, but will be in the future. In the meantime, you can use the
+    /// third-party [tablex library](https://github.com/PgBiel/typst-tablex/).
+    #[resolve]
+    #[fold]
+    pub stroke: Option<Stroke>,
+
+    /// How much to pad the cells' content.
+    ///
+    /// ```example
+    /// #grid(
+    ///   inset: 10pt,
+    ///   fill: (_, row) => (red, blue).at(row),
+    ///   [Hello],
+    ///   [World],
+    /// )
+    ///
+    /// #grid(
+    ///   columns: 2,
+    ///   inset: (
+    ///     x: 20pt,
+    ///     y: 10pt,
+    ///   ),
+    ///   fill: (col, _) => (red, blue).at(col),
+    ///   [Hello],
+    ///   [World],
+    /// )
+    /// ```
+    #[fold]
+    #[default(Sides::splat(Abs::pt(0.0).into()))]
+    pub inset: Sides<Option<Rel<Length>>>,
```

Comment:
Should we make `inset` Celled in this PR? Or leave this for later maybe (perhaps to make it faster to review this PR, since this isn't urgent)?

Reply:
If it's a small diff with no further complication, I'm fine with having it in this PR.

Reply:
Right... I'm having some trouble because of `#[fold]`. `Celled<T>` doesn't implement `Fold`, so I tried to implement `Fold` for it using something like

```rs
impl<T: Fold<Output = T>> Fold for Celled<T> {
    type Output = Self;

    fn fold(self, outer: Self::Output) -> Self::Output {
        match (self, outer) {
            (Self::Value(inner), Self::Value(outer)) => Self::Value(inner.fold(outer)),
            (inner, _) => inner
        }
    }
}
```

Note the `T: Fold<Output = T>` bound there; that was necessary for `Output` to be `Self`. But that doesn't work, because `Sides<Option<T>>::Output` is `Sides<T>`, so `<Celled<T> as Fold>::Output` must be `Celled<T::Output>` instead. So I tried something like

```rs
impl<T: Fold> Fold for Celled<T> {
    type Output = Celled<T::Output>;

    fn fold(self, outer: Self::Output) -> Self::Output {
        match (self, outer) {
            (Self::Value(inner), Celled::Value(outer)) => Celled::Value(inner.fold(outer)),
            (inner, _) => inner,
        }
    }
}
```

but that doesn't work because `inner` is `Celled<T>`, not `Celled<T::Output>`.

So I think we'd have to give up on fold for `inset` if we want this to work properly... though it would be nice to have fold for the `(Value, Value)` case at least... 🤔

Anyway... maybe we can just leave this to another PR or something so we don't have to think about this now :p

Reply:
Maybe that's the reason I didn't add it in the first place!